include/smack/smack.h

```diff
@@ -135,32 +224,143 @@ void __SMACK_decls() {
   D("axiom (forall i: int :: $fp2si($si2fp(i)) == i);");
   D("axiom (forall f: float :: $si2fp($fp2si(f)) == f);");
 
-  // Memory Model
-  D("function $base(int) returns (int);");
+#ifdef BITVECTOR
+  D("type i8 = bv8;");
+  D("type i16 = bv16;");
+  D("type i32 = bv32;");
+  D("type i64 = bv32;");
+  D("type ref = i64;");
+  D("type size_t = i32;");
+#else
+  D("type i8 = int;");
+  D("type i16 = int;");
+  D("type i32 = int;");
+  D("type i64 = int;");
+  D("type ref = int;");
+  D("type size_t = int;");
+#endif
 
-  D("const unique $NULL: int;");
-  D("axiom $NULL == 0;");
+  // Memory Model
   D("const $UNDEF: int;");
+  D("function $base(ref) returns (ref);");
+  D("const unique $NULL: ref;");
+#ifdef BITVECTOR
+  D("axiom $NULL == 0bv32;");
+  D("function {:inline} $pa(pointer: ref, index: ref, size: ref) returns (ref) {$add.i32(pointer, $mul.i32(index, size))}");
```

Comment:
Yes, I think I can. 

LLVM Language Reference Manual states, 
>The type of each index argument depends on the type it is indexing into. When indexing into >a (optionally packed) structure, only i32 integer constants are allowed (when using a vector >of indices they must all be the same i32 integer constant). When indexing into an array, >pointer or vector, integers of any width are allowed, and they are not required to be constant. >These integers are treated as signed values where relevant.

Based on my experiment, it seems index of structure types is `i32` while that of other types is `i64`. I guess the latter is probably platform-specific.